### PR TITLE
[LLVM][IR] Add textual shorthand for specifying constant vector splats.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
Add LL parsing for `<N x ty> splat(ty <imm>)` that lowers onto ConstantInt::get() for integer types and ConstantFP::get() for floating-point types.

The intent is to extend ConstantInt/FP classes to support vector types rather than redirecting to other constant classes as the get() methods do today.

This patch gives IR writers the convenience of using the shorthand today, thus allowing existing tests to be ported.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -4299,6 +4299,11 @@ constants and smaller complex constants.
     "``< i32 42, i32 11, i32 74, i32 100 >``". Vector constants
     must have :ref:`vector type <t_vector>`, and the number and types of
     elements must match those specified by the type.
+
+    When creating a vector whose elements have the same constant value, the
+    prefered syntax is ``splat (<Ty> Val)``. For example: "``splat (i32 11)``".
```
Comment (paulwalker-arm) [inline llvm/docs/LangRef.rst:4304]: Am I being too adventurous here by suggesting this be the preferred syntax?

**Comment by nikic** [comment]
My main question here is whether this should be limited to immediates, or work with everything that `ConstantVector::getSplat()` supports. That is, is there any reason why this syntax supports `splat (i32 123)` but not for example `splat (ptr @g)` or `splat (i64 ptrtoint (ptr @g to i64))`?

Ultimately we will need some way to represent all splats we can currently represent -- I assume that the `splat (ptr @g)` case for scalable vectors is going to be covered by `getelementptr i8, ptr @g, <vscale x N x i64> zeroinitializer`. (Are there any other tricky cases?)

**Comment by paulwalker-arm** [comment]
It depends how far down the rabbit hole we want to go.  From a textual point of view the change looks easy but given the goal is to move away from `ConstantExpr` I'd have to ask what the value is to supporting `splat (ptr @g)` for scalable vectors?  We don't support scalable vector globals so the only use is within a function where non-constant IR should be sufficient? I'm generally not that familiar with the benefits and the existing scalable vector support comes from it being a logical extension of existing code rather than having a massive amount of thought behind it.

For today perhaps it's best I extend the textual support to fallback to `ConstantVector::getSplat()` rather than erroring and defer the decision whether to deprecate non-immediate scalable vector constants.  My guess is it'll take some time to migrate the immediate cases anyway.

**Comment by nikic** [comment]
> It depends how far down the rabbit hole we want to go. From a textual point of view the change looks easy but given the goal is to move away from `ConstantExpr` I'd have to ask what the value is to supporting `splat (ptr @g)` for scalable vectors? We don't support scalable vector globals so the only use is within a function where non-constant IR should be sufficient? I'm generally not that familiar with the benefits and the existing scalable vector support comes from it being a logical extension of existing code rather than having a massive amount of thought behind it.

My general thought here was that this is necessary for "closure", so that the `ConstantVector::getSplat()` can safely be used with any `Constant *` and we don't get edge case assertion failures where scalable vectors are combined with constant expressions. But after looking at some of the uses of `ConstantVector::getSplat()`, I'm coming around to the idea that we probably don't need this.

File: llvm/docs/LangRef.rst (hunk 1 above)
Comment (nikic) [inline llvm/docs/LangRef.rst:4304]:
```suggestion
    preferred syntax is ``splat (<Ty> Val)``. For example: "``splat (i32 11)``".
```

File: llvm/lib/AsmParser/LLParser.cpp [hunk 2]
```diff
@@ -3981,6 +3981,33 @@ bool LLParser::parseValID(ValID &ID, PerFunctionState *PFS, Type *ExpectedTy) {
...
+    Lex.Lex();
+    if (parseToken(lltok::lparen, "expected '(' after vector splat"))
+      return true;
+    Constant *C;
+    if (parseGlobalTypeAndValue(C))
+      return true;
+    if (parseToken(lltok::rparen, "expected ')' at end of vector splat"))
+      return true;
+
+    if (auto *CI = dyn_cast<ConstantInt>(C)) {
```
Comment (nikic) [inline llvm/lib/AsmParser/LLParser.cpp:3994]: Why is the special ConstantInt/ConstantFP handling here needed?

File: llvm/lib/AsmParser/LLParser.cpp [hunk 3]
```diff
@@ -5740,14 +5767,15 @@ bool LLParser::convertValIDToValue(Type *Ty, ValID &ID, Value *&V,
       V = NoCFIValue::get(cast<GlobalValue>(V));
     return V == nullptr;
   case ValID::t_APSInt:
-    if (!Ty->isIntegerTy())
+    if (!Ty->isIntOrIntVectorTy())
```
Comment (nikic) [inline llvm/lib/AsmParser/LLParser.cpp:5770]: I don't really get why these changes are needed. Isn't the value parsing itself still performed with the scalar type?

File: llvm/lib/AsmParser/LLParser.cpp [hunk 4]
```diff
@@ -3981,6 +3981,33 @@ bool LLParser::parseValID(ValID &ID, PerFunctionState *PFS, Type *ExpectedTy) {
...
+    }
+
+    if (auto *CFP = dyn_cast<ConstantFP>(C)) {
+      ID.APFloatVal = CFP->getValue();
+      ID.Kind = ValID::t_APFloat;
+      return false;
+    }
+
+    ID.ConstantVal = C;
+    ID.Kind = ValID::t_Constant;
```
Comment (nikic) [inline llvm/lib/AsmParser/LLParser.cpp:4007]: I would expect the `ConstantVector::getSplat()` call to occur here, rather than in the t_Constant branch. If it's done there, I think it might hide vector vs scalar mismatches?

File: llvm/lib/AsmParser/LLParser.cpp (hunk 2 above)
Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:3994]: Constant values are parsed separately from the type of the resulting expression, with the `ValID` union used as a proxy to hold the parsed value.

File: llvm/lib/AsmParser/LLParser.cpp (hunk 4 above)
Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:4007]: As above.  Constants are parsed, stashed in a `ValID` union and the various Constant classes are create afterwords once the type is known.

File: llvm/lib/AsmParser/LLParser.cpp (hunk 3 above)
Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:5770]: This boils down to the same reason as above.  Floating point looks to be handled a little stricter whereas integers just resize the `ID.APSIntVal`. I've no knowledge as to the reason for this so have just maintained the current approach.

File: llvm/lib/AsmParser/LLParser.cpp (hunk 2 above)
Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:3994]: It turns out when extending the work to cover non-immediate values I inadvertently introduced a parsing bug so I don't think I can safely use `t_Constant`.  That said, this doesn't effect the core of your suggestion so I'll funnel all the splat cases down the same path, I'll just need to give it a new name.

File: llvm/lib/AsmParser/LLParser.cpp [hunk 5]
```diff
@@ -3981,6 +3981,21 @@ bool LLParser::parseValID(ValID &ID, PerFunctionState *PFS, Type *ExpectedTy) {
...
+    if (parseToken(lltok::lparen, "expected '(' after vector splat"))
+      return true;
+    Constant *C;
+    if (parseGlobalTypeAndValue(C))
+      return true;
+    if (parseToken(lltok::rparen, "expected ')' at end of vector splat"))
+      return true;
+
+    ID.ConstantVal = C;
+    ID.Kind = ValID::t_ConstantSplat;
```
[the remark this answers is not in the document]
Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:3995]: We don’t have the target type at this point in the parsing (NOTE `ExpectedTy` defaults to `nullptr` and is only used in one part of `parseValID`).  ConstantExprs can be constructed based on their input operands, with only their final result type needing to be validated within `convertValIDToValue`. 

The reason for not reusing `t_Constant` is to maintain the accuracy of the diagnostics.

[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/AsmParser/LLParser.cpp:3995]: > We don’t have the target type at this point in the parsing (NOTE `ExpectedTy` defaults to `nullptr` and is only used in one part of `parseValID`).

Ah, thanks, this is what I was missing.

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/lib/AsmParser/LLParser.cpp [hunk 6]
```diff
@@ -5824,6 +5839,17 @@ bool LLParser::convertValIDToValue(Type *Ty, ValID &ID, Value *&V,
                                "' but expected '" + getTypeString(Ty) + "'");
     V = ID.ConstantVal;
     return false;
+  case ValID::t_ConstantSplat:
+    if (!Ty->isVectorTy())
+      return error(ID.Loc, "vector constant must have vector type");
```
Comment (nikic) [inline llvm/lib/AsmParser/LLParser.cpp:5844]: If you feel like it, we could test these error conditions, but I'm not going to insist...

Comment (paulwalker-arm) [reply llvm/lib/AsmParser/LLParser.cpp:5844]: Sure, I'll add them.  If more existed I would not have made my original mistake :)
